### PR TITLE
Thomas/heuristic pipeline

### PR DESCRIPTION
**Pull request by Mostah**

### CODE REVIEW

File: src/learning/HeuristicPipeline.cpp [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+  const auto before_init = clock::now();
+  for (int k = 0; k < conf.model_batch_size; k++) {
+    MRSortModel model = MRSortModel(n_cat, n_crit);
+
+    profileInitializer.initializeProfiles(model);
+    // change back to alt mode
+    model.profiles.changeMode("alt");
+    this->computeAccuracy(model);
+    std::cout << "accuracy after init: " << model.getScore() << std::endl;
```
Comment (thibmonsel) [inline src/learning/HeuristicPipeline.cpp]: maybe put this next to accuracy printing for LP solving in order to get the evolution of the accuracy in order ?

File: src/learning/HeuristicPipeline.cpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  }
+  const sec weight_duration = clock::now() - before_weight;
+  const auto before_profile = clock::now();
+  // Update profiles
+  for (int k = 0; k < conf.model_batch_size; k++) {
+    profileUpdater.updateProfiles(models[k]);
+    this->computeAccuracy(models[k]);
+    std::cout << "accuracy after profile update: " << models[k].getScore()
+              << std::endl;
+  }
```
Comment (thibmonsel) [inline src/learning/HeuristicPipeline.cpp]: according to olivier but we would have to re-check his python, for one LP he did several Prof updating

File: src/learning/HeuristicPipeline.cpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    // models are sorted in descending order by getScore()
+    // re-initialize the worst half of the models
+    for (int k = conf.model_batch_size - 1; k > conf.model_batch_size / 2 - 1;
+         k--) {
+      models[k].criteria.generateRandomCriteriaWeights();
+      profileInitializer.initializeProfiles(models[k]);
+      models[k].profiles.changeMode("alt");
+      this->computeAccuracy(models[k]);
+      std::cout << "accuracy after re-init: " << models[k].getScore()
+                << std::endl;
```
Comment (thibmonsel) [inline src/learning/HeuristicPipeline.cpp]: same goes here

Comment (thibmonsel) [reply src/learning/HeuristicPipeline.cpp]: you can print the accuracy at the start of the LP for loop before executing anything

File: src/types/DataGenerator.cpp [hunk 4]
```diff
@@ -256,7 +256,7 @@ void DataGenerator::saveModel(std::string fileName, float lambda,
   int nb_categories = pt.getPerformanceTable().size();
   int nb_criteria = criteria.getCriterionVect().size();
 
-  if (nb_categories != pt.getPerformanceTable()[0].size()) {
+  if (nb_criteria != pt.getPerformanceTable()[0].size()) {
```
Comment (thibmonsel) [inline src/types/DataGenerator.cpp:259]: nice one !

**Review by thibmonsel — APPROVED** [review]
Looks good, put small comments that might be interesting to look at.

File: src/app.cpp [hunk 5]
```diff
@@ -3,44 +3,57 @@
...
-  std::cerr
-      << "Usage: " << name << " <option(s)> SOURCES"
-      << "Options:\n"
-      << "\t-h,--help\t\tShow this help message\n"
-      << "\t-e,--env ENVIRONMENT\tSpecify the environment: staging or prod"
-      << std::endl;
+  std::cerr << "Usage: " << name << " <option(s)> SOURCES"
+            << "Options:\n"
+            << "\t-h,--help\t\tShow this help message\n"
+            << "\t-d,--dataset DATASET\tDataset path\n"
```
Comment (edix9) [inline src/app.cpp]: maybe add "Dataset FILE path" and "Output FILE path" to ensure they don't enter a directory?

File: src/learning/HeuristicPipeline.cpp [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    std::ostringstream ss;
+    ss << "accuracy of model " << k
+       << " after profile update: " << models[k].getScore() << std::endl;
+    conf.logger->debug(ss.str());
+  }
+  const sec profile_duration = clock::now() - before_profile;
+
+  std::ostringstream ss0;
+  ss0 << "Profile initialization of all models took: " << init_duration.count()
+      << "s" << std::endl;
```
Comment (edix9) [inline src/learning/HeuristicPipeline.cpp]: should we add a % of the total time?

File: src/learning/HeuristicPipeline.cpp [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+      this->computeAccuracy(models[k]);
+
+      std::ostringstream ss;
+      ss << "accuracy of model " << k
+         << " after profile update: " << models[k].getScore() << std::endl;
+      conf.logger->debug(ss.str());
+    }
+    this->orderModels();
+    conf.logger->info("Iteration " + std::to_string(i) +
+                      " done, best model has a score of: " +
```
Comment (edix9) [inline src/learning/HeuristicPipeline.cpp:137]: -> best model (k) _(id model)_ and has a score of .. ?

**Review by edix9 — APPROVED** [review]
good job!

File: src/learning/HeuristicPipeline.cpp (hunk 6 above)
Comment (Mostah) [reply src/learning/HeuristicPipeline.cpp]: Yep, I'll add that 👍

**Pull request merged**